File: java/sorting/merge-sort.java
class MergeSort {
    public static void main(String[] args) {
        int[] a = {43,76,22,543,6,2,33,3,7,4,5,8,9,0,54,66,23,77,78,34,87,23,42,224,76575,45,777,556,64};
        printArray(a);
        sort(a, 0, a.length - 1);
        printArray(a);
    }

    public static void merge(int[] arr, int i, int j, int m) {
        int l = i;
        int r = m + 1;
        int[] tempArr = new int[(j-i) + 1];
        int count = 0;

        while(l <= m && r <= j) {
            if(arr[l] <= arr[r]) {
                tempArr[count] = arr[l];
                l = l + 1;
                count = count + 1;
            }
            else {
                tempArr[count] = arr[r];
                r = r + 1;
                count = count + 1;
            }
        }

        while(l <= m) {
            tempArr[count] = arr[l];
            l = l + 1;
            count = count + 1;
        }

        while(r <= j) {
            tempArr[count] = arr[r];
            r = r + 1;
            count = count + 1;
        }

        int k = i;
        while(k <= j) {
            arr[k] = tempArr[k-i];
            k = k + 1;
        }
    }

    public static void sort(int[] arr, int i, int j) {
        if(j > i) {
            int mitad = ((j-i)/2) + i;
            sort(arr, i, mitad);
            sort(arr, mitad + 1, j);
            merge(arr, i, j, mitad);
        }
    }

    public static void printArray(int[] arr){
        System.out.print("[");
        for(int i = 0; i < arr.length; i++){
            if(i == arr.length - 1){
                System.out.print(arr[i]);
                break;
            }
            System.out.print(arr[i] + ", ");
        }
        System.out.println("]");
    }
}

File: zig/sorting/merge-sort.zig
const std = @import("std");

fn merge(arr: []i32, i: usize, j: usize, m: usize, arena: *std.heap.ArenaAllocator) !void {
    var l: usize = i;
    var r: usize = m + 1;
    var tempArr = try arena.allocator().alloc(i32, (j - i) + 1);
    defer arena.allocator().free(tempArr);
    var count: usize = 0;

    while (l <= m and r <= j) : (count = count + 1) {
        if (arr[l] <= arr[r]) {
            tempArr[count] = arr[l];
            l = l + 1;
        } else {
            tempArr[count] = arr[r];
            r = r + 1;
        }
    }

    while (l <= m) : (count = count + 1) {
        tempArr[count] = arr[l];
        l = l + 1;
    }

    while (r <= j) : (count = count + 1) {
        tempArr[count] = arr[r];
        r = r + 1;
    }

    var k: usize = i;
    while (k <= j) : (k = k + 1) {
        arr[k] = tempArr[k - i];
    }
}

fn sort(arr: []i32, i: usize, j: usize, arena: *std.heap.ArenaAllocator) !void {
    if (j > i) {
        const mitad: usize = @divTrunc(j - i, 2) + i;
        try sort(arr, i, mitad, arena);
        try sort(arr, mitad + 1, j, arena);
        try merge(arr, i, j, mitad, arena);
    }
}

pub fn main() !void {
    var a = [_]i32{ 43, 76, 22, 543, 6, 2, 33, 3, 7, 4, 5, 8, 9, 0, 54, 66, 23, 77, 78, 34, 87, 23, 42, 224, 76575, 45, 777, 556, 64 };
    std.debug.print("{any}\n", .{a});
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    try sort(&a, 0, a.len - 1, &arena);
    std.debug.print("{any}\n", .{a});
}
